contracts/extensions/OneTxPayment.sol

```diff
@@ -181,100 +149,86 @@ contract OneTxPayment is ColonyExtension {
-      "colony-one-tx-payment-arrays-must-be-equal-length"
+      colony.hasInheritedUserRole(msg.sender, _callerPermissionDomainId, FUNDING, _callerChildSkillIndex, _domainId) &&
+      colony.hasInheritedUserRole(msg.sender, _callerPermissionDomainId, ADMINISTRATION, _callerChildSkillIndex, _domainId),
+      "one-tx-payment-not-authorized"
     );
-    // Get around stack too deep with array
-    // paymentData[0] = expenditure slot
-    // paymentData[1] = expenditureId/paymentId
-    // paymentData[3] = fundingPotId
-    uint256[3] memory paymentData;
-    paymentData[0] = 0;
-    ColonyDataTypes.Domain memory domain = colony.getDomain(_domainId);
-    if (_tokens.length > 1) {
-      // Make a new expenditure
-      paymentData[1] = colony.makeExpenditure(_permissionDomainId, _childSkillIndex, _domainId);
-      ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(paymentData[1]);
-      paymentData[2] = expenditure.fundingPotId;
-      colony.setExpenditureRecipient(paymentData[1], paymentData[0], _workers[0]);
-      colony.setExpenditureSkill(paymentData[1], paymentData[0], _skillId);
+
+    if (_workers.length == 1) {
+
+      uint256 paymentId = colony.addPayment(_permissionDomainId, _childSkillIndex, _workers[0], _tokens[0], _amounts[0], _domainId, _skillId);
+      uint256 fundingPotId = colony.getPayment(paymentId).fundingPotId;
+      uint256 domainPotId = colony.getDomain(_domainId).fundingPotId;
+
+      moveFundsWithinDomain(_permissionDomainId, _childSkillIndex, domainPotId, fundingPotId, _amounts[0], _tokens[0]);
+
+      colony.finalizePayment(_permissionDomainId, _childSkillIndex, paymentId);
+      colony.claimPayment(paymentId, _tokens[0]);
+
     } else {
-      // Add a new payment
-      paymentData[1] = colony.addPayment(
-        _permissionDomainId,
-        _childSkillIndex,
-        _workers[0],
-        _tokens[0],
-        _amounts[0],
-        _domainId,
-        _skillId);
-      ColonyDataTypes.Payment memory payment = colony.getPayment(paymentData[1]);
-      paymentData[2] = payment.fundingPotId;
-    }
-    for (uint256 index = 0; index < _workers.length; index++) {
-      if (_tokens.length > 1) {
-        if (index != 0 && _workers[index] != _workers[index-1]) {
-          paymentData[0]++;
-          colony.setExpenditureRecipient(paymentData[1], paymentData[0], _workers[index]);
-          colony.setExpenditureSkill(paymentData[1], paymentData[0], _skillId);
+
+      uint256 expenditureId = colony.makeExpenditure(_permissionDomainId, _childSkillIndex, _domainId);
+      uint256 fundingPotId = colony.getExpenditure(expenditureId).fundingPotId;
+      uint256 domainPotId = colony.getDomain(_domainId).fundingPotId;
+
+      uint256 idx;
+      uint256 slot;
+
+      for (idx = 0; idx < _workers.length; idx++) {
+        moveFundsWithinDomain(_permissionDomainId, _childSkillIndex, domainPotId, fundingPotId, _amounts[idx], _tokens[idx]);
```

Comment:
Mm... my only reservation would have been gas costs and loops, but ultimately it doesn't really matter if this transaction hits a gas limit and I expect that in most cases this will be gas-saving anyway. I'll think if I can figure out a smart way to do this.